### ISSUE TITLE
Discussion: "std" package naming

### ISSUE THREAD

**Issue by leohhhn** [gnolang/gno#1475]
Opening the discussion on the naming of the `"std"` package within Gno. 

While writing the docs, I came across an issue on how to differentiate the Standard libraries (ie. `"encode/binary"`, `"strings"`, `"testing"` etc), and the `"std"` package itself, which contains blockchain-specific types and functionalities, like `std.Address`, `std.GetBanker()`, `std.GetOrigCaller()` etc.

Have we considered renaming this package to avoid confusion? Some of the ideas are as follows:
- `"os"` - since `os` isn't used, and `std` gives information about the chain/vm state, this could make sense,
- `"chain"` - makes sense to have chain-related things called by `chain.GetHeight()`, `chain.GetBanker()`, etc,
- or simply have them imported by default, ie you can call `GetBanker()` without the `.` notation. 

What do you guys think?

@moul @thehowl @zivkovicmilos

**Comment by waymobetta** [comment]
I'm with the opinion to avoid overwriting any existing Go stdlib package names to avoid confusing those who often switch between the languages, searchability within the codebase (as `os` is used internally), as well as misc. auto-import confusions.

I don't think renaming this library is necessary but I'm curious to hear what others think; or we can determine as a collective rather how to refer to the current `std` package, as we have been discussing.

**Comment by moul** [comment]
I have experienced similar feelings on multiple occasions.

Here are my current ideas, listed randomly:
-  Keep the name as `std`.
-  Change the name to `os` to align with Golang's package. However, this would result in significant API differences. For a blockchain-less version of gnovm, such as for CLI scripting, we could enable the actual `os`.
-  Rename it to `gno` since it is specific to running on a gnovm blockchain. We can imagine a blockchain-less version of a gnovm-powered server with a completely different `std` package.
-  Move it to a userland package like `gno.land/p/std` to only have shared std libraries with Go. This would be facilitated by merging the native binding PR (#859), but it might create more ambiguity.

**Comment by leohhhn** [comment]
@moul 

With 1. we will need workarounds in order to discuss/document the std package (not that big of a deal, but still), so essentially I am for modifying the name.

I believe 2. is a totally valid concern. I think `"gno"` is quite intuitive and unique enough to not get confused with anything. 

Are there any challenges in regarding to doing this kind of renaming?

**Comment by moul** [comment]
I can't see any significant technical challenges. It's primarily a challenge related to language design and concerns about adoption and documentation.

Remember that it will be the most widely used import. It's kind of similar with choosing the right keyword, like "return," "panic," "switch," or "for,". It should be something memorable, similar to the well-known `if err != nil`, which is strongly associated with go.

**Comment by deelawn** [comment]
I like the name `std` and not renaming it to `os` or something else that overloads an existing go standard library package name.  In my opinion, `std` is the perfect name to avoid naming conflicts with go standard library packages because go does not allow packages to be named `std`.  If we want to apply more meaningful names to groups of functionality within `std`, then I'd suggest grouping them into subpackages.

I think it is better to keep it out of the normal "userland" package space because of the native bindings that it is built on.  I think it makes sense in general for ONLY standard library packages to be able to define native bindings.  And in fact, now that the native bindings PR has been merged, it ensures that they are only supported in standard library packages.

So TLDR; I don't support renaming this package but am fine with reorganizing into subpackages.

**Comment by thehowl** [comment]
I largely second others in thinking that putting `std` functionality in `os` is a tough fit. I think `time.Now()` returning the block time is a good exception, because as long as we explain the differences, everything else in time still works the same (though, even then, a good consideration might be to rename `Now()` to something more descriptive like `BlockTime()`, but I digress).

I don't like `std` as a name because I think it is too generic and would like for it to be broken down. `os` is largely not a good fit and a misnomer (you are not accessing the OS, and calling Gno.land an "OS" is a stretch of imagination). But I think `runtime`, in this regard, is a much better fit. Go's `runtime` package already exposes many meta-runtime features that make sense within Go programs, so I think in this case we can have the `runtime` package with a different API.

I propose the following (sorry, I detoured on a redesign of all of the function names, hope you'll forgive me):

- Package `runtime/chain` OR just `chain` (or variants `blockchain`, `bc`), containing:
  - types `Address`, `AddressList`
  - `DerivePkgAddr` (better name: `AddressFromPkgPath`?)
  - `EncodeBech32`/`DecodeBech32`
  - type `Banker`
  - type `Coin`, `Coins` 
  - type `Realm`
  - type `RawAddress` (better: either remove, or make functions like `EncodeBech32/DecodeBech32` use it
- Package `runtime`, containing:
  -  `GetCallerAt` (better: `CallerAt`)
  - `GetOrigCaller` (better: `OriginCaller`)
  - `GetOrigPkgAddr` (better: axe this in favour of `OriginPkgPath` + `chain.AddressFromPkgPath`)
  - `GetBanker` (better: `NewBanker`) + `BankerType`
  - `GetOrigSend` (better: `OriginSend`)
  - `CurrentRealm`, `PrevRealm`
  - `CurrentRealmPath` (better: axe in favour of CurrentRealm.Path)
  - `GetChainID` (better: `ChainID`)
  - `GetHeight` (better: `Height`, `BlockHeight`)
  - `AssertOriginCall` / `IsOriginCall`

In other words, I like to have a package `chain` which defines key blockchain-specific types (and utility functions for them), and have a `runtime` package that accesses the "meta-gnovm" features.

In a kind of way, this is similar to package `io` (define types, interfaces and useful helper functions) and package `os` (access the filesystem, often through `io` types).

**Comment by thehowl** [comment]
Also, one more thought.

I'd reserve `gno` to be used as a keyword, just-in-case. (And thus not use it as an identifier anywhere)

Maybe the way we implement goroutines is different enough that a better fit is to change the keyword entirely to `gno` :)

**Comment by ilgooz** [comment]
`os/gno`?

Similar to [syscall/js](https://pkg.go.dev/syscall/js)

**Comment by leohhhn** [comment]
Let this comment be a poll:

Idea 1: as per @thehowl's [comment](https://github.com/gnolang/gno/issues/1475#issuecomment-1866110827), split `std` into `chain` & `runtime`. Vote with 🚀 

Idea 2: simply rename `std` to `gno`. Vote with  🎉 

Idea 3: keep `std` as is. Vote with ❤️ 

In addition, I think that no matter the case, we should decouple classic `std` functions from functions used for test context editing (ie `std.TestSetOrigCaller`), and do some name refactoring to make function names more clear from the get-go.

**Comment by grepsuzette** [comment]
I simply advocate renaming `std` to `gno` package.

I do agree splitting packages *makes* sense as possible solution, however people are not only going to program exclusively in gno in the future. 

`import "gno"` is extremely simple, it's as simple and automatic no-brainer as `import "std"` today.  From there, people get autocompletion upon typing `gno.|`. It doesn't get more simple for beginners. The number of functions in `std` is low enough. 

Beginners and nobody else are the ones who will become long-term users, and for that they need a simple language.

### About reserving `gno` as a future keyword

The fact `go func(){ .. }()` exists does not forces us to reserve `gno func() { .. }()`.
If we need something special (don't deny, it's a possibility), we can always work a longer name like `gnoroutine`. 

### We should choose the most impactful 

*gno as a keyword* is never not going to be as impactful as *gno as a package*.

**Comment by thehowl** [comment]
> I simply advocate renaming `std` to `gno` package.

I appreciate the "natural feeling" given by typing out `gno.Address`, however I do think we should use the term "gno", including in our standard libraries, to refer to the programming language rather than the chain.

This is similar to what happens in go; where the `go/*` packages are those concerning the Go programming language (ie. parsing, AST, type checker...).

> I do agree splitting packages _makes_ sense as possible solution[1](#user-content-fn-1-d0406d0a446de18ffbd0783792fdb854), however people are not only going to program exclusively in gno in the future.

What?

People aren't going to programming exclusively for gno.land in the future. But this is a change that concerns primarily the standard libraries, which are likely to be brought over into other gno.land ecosystems. This actually strengthens my argument; that `chain` translates better when using gno elsewhere than having a `gno`  package.

> Beginners (in gno) and nobody else are the ones who will become long-term users, and for that conversion rate to be maximized I think they need a simple language.
> 
> `import "gno"` is extremely simple, it's as simple and automatic no-brainer as `import "std"` today. From there, people get autocompletion upon typing `gno.|`. It doesn't get more simple for beginners.

good rhetoric ;)

i don't think we'll necessarily end up using it as a keyword; it's a backward-incompatible change with go, i'd much rather use `go`.

but from principles, i disagree that `gno` is a better way to namespace types like `Address`, `Coins` and `Banker`. They relate to the blockchain it's referencing, not the programming language.

**Comment by Kouteki** [comment]
Related issues, to discuss as a single solution

- https://github.com/gnolang/gno/issues/1576
- https://github.com/gnolang/gno/issues/1577

**Comment by moul** [comment]
Tentative reorganization: https://gist.github.com/moul/f28f4b12864fd40a2ca6b6c20294da27

**Comment by thehowl** [comment]
Combining all the proposals into the top-level comment of this issue

**Comment by thehowl** [comment]
Updated the [top-level comment](https://github.com/gnolang/gno/issues/1475#issue-2050626558) to centralize all proposals.

@moul Please take a look; I tried to summarize also the changes that I tried making with #2425 so that they can be tackled by someone other than me, as we talked about.

What do you think about having a call to go through all of these and figuring out what proposals are sensible and to act on?

I think we can work on these in small PRs, but possibly do all breaking changes ahead of test6.

---

As a side note, I added the `PkgPath` / `ImportPath` problem as well as we talked about it and I think it's still worth a discussion, but I'm more undecided after reading @moul's comment, and especially after I discovered, to my disbelief, that [Go in some cases uses PkgPath as well](https://pkg.go.dev/golang.org/x/tools/go/packages#Package.PkgPath).

**Comment by moul** [comment]
Let's schedule a call. Initially, I expected to create a PR instead of this issue so we could comment line by line. However, the problem is that each proposal will not only change lines but also file locations. Therefore, let's have a call.

Regarding whether to create a single PR or multiple ones, my main concern is fixing the history of the portal loop. This depends on #3024 as evidence that we can patch the history for such breaking changes.

Let's assume it will be ready, but we rely not only on developers but also on DevOps to ensure the portal loop loads the Git state. Alternatively, if @gfanton  can make gnodev the next portal loop with this feature, that could also work.

**Comment by thehowl** [comment]
Thursday called; reached full consensus on the first three proposals. For the fourth, "Remove std.GetOrigCaller and std.IsOriginCall", let's come back at it next week with more usage examples (both in favour and against the inclusion of the two functions).

**Comment by piux2** [comment]
Here are my two cents:

### `IsOriginCall`
We don't need `IsOriginCall`; instead, we should use `AssertOriginCall`, as it serves the purpose of eliminating the middleman and removes the risk of not handling errors properly.

### `GetOrginCaller `
We do need to keep `GetOriginCaller`.
 
It is used to ensure that the address returned is from an end user's wallet, also known as an external account. The most important security property of GetOriginCaller is that the original message's signature is verified by the chain against GetOriginCaller.

For all non-external callers, such as GetPrevRealm, where the caller is a realm, the contract must trust the caller realm. In use cases where a middle contract needs to be involved, GetOriginCaller acts as a safeguard, delegating trust to the chain rather than the caller contract.

There are two categories of use cases where we need to verify the external account while allowing a middleman contract.. 


### A. Asset  (GRC20, NFT) Issued Directly to End Users, with Exchanges or Aggregators Involved

Assets are either directly issued to an end user's wallet or claimed by the end user, who then places them on an exchange or auction house for trading. Aggregators, exchanges, and auction houses are considered middleman contracts.

Example flow:

Original Caller (External Account) → Exchange Aggregator (Third Party A) → Token Exchange (Third Party B) → GRC20 (Token Issuer)


For a safe token withdrawal, exchange contracts need to use GetOriginCaller to ensure that tokens return to the address originally stored in the GRC20 contract, where tokens were directly issued to the end user's wallet address (External Account). In other words, it ensures that the transaction is initiated from a user's wallet.


### B. Asset Issuers Distributing Assets (GRC20, NFTs, Insurance Policies) via Brokers

Example flow:

Original Caller (External Account) → NFT Broker (Third Party Entity) → GRC721 (NFT Issuer)

NFTs are distributed through a third party as a reward to end users (External Account). The NFT issuer contract needs to verify that the NFT is claimed from an external account.


### C: Permissioned Use Cases Combining GetOriginCaller and GetPrevRealm

For permissioned scenarios, we can combine GetOriginCaller and GetPrevRealm to whitelist middleman contracts. GetOriginCaller ensures that the transaction is initiated by an end user, while GetPrevRealm ensures the transaction is from a whitelisted middleman.

@leohhhn The issue you mentioned with other chains is a bit different because they don't have AssertOriginCall built into the language to eliminate the middleman contract. They only have something similar to GetOriginCaller and GetPrevRealm. For contracts that actually do verify direct withdrawals from external accounts, people have to write something like GetOriginCaller == GetPrevRealm in their contracts.  

## Other thoughts.

I see the complication arises from GetPrevRealm returning the same information as GetOriginCaller in direct call cases.

In abstract terms, we can consider a user's external account as a realm, especially when MsgRun is involved, in which the external account signs and executes a temporary realm contract called "main."

However, in reality, GetPrevRealm returns user information that overlaps with other functions, and it cannot replace GetOriginCaller or AssertOriginCall.

When verifying the OriginCaller, there are a few design options:

- Eliminate the Middleman

Option 1: `GetOriginCaller + AssertOriginCall`
Option 2: `PrevRealm.IsUser + PrevRealm.Address + AssertOriginCall`
Option 3: `PrevRealm.IsUser + GetOriginCaller + other logic to handle the negative case`
Option 4: `PrevRealm.IsUser + PrevRealm.Address+ other logic to handle the negative case`

- Whitelist the Middleman

`GetOriginCaller, PrevRealm.IsRealm, PrevRealm.Address`

- Permissionless Middleman

GetOriginCaller

As we can see, PrevRealm.IsUser is only useful in the first case, where GetOriginCaller + AssertOriginCall is the cleaner option.

What if we remove the mental burden that PrevRealm could also return a user address? Instead, PrevRealm either returns a realm on-chain or it doesn't.

This way, everything is clean and straightforward:

- Eliminate the Middleman

`GetOriginCaller` + `AssertOriginCall`

- Whitelist the Middleman

`GetOriginCaller`, `PrevRealm.OK`,` PrevRealm.Address`

- Permissionless Middleman

`GetOriginCaller`

**Comment by thehowl** [comment]
Updated the OP to re-arrange many of the in progress items into their own section.

As it currently stands, these two proposals are to be implemented next:

- Split into different packages
- Manipulate entire contexts instead of `TestSetX` functions (no approval from Jae, but it concerns the testing context so it is easy to change post-launch)

Ahead of the mainnet launch, the other changes I'd like to see happen and get approval, hopefully this thursday, are:

- Removing CallerAt, GetOrigPkgAddr, RawAddress
- Realm-scoped Deposits instead of OrigSend
- Banker changes

The last two changes, "Something other than Realm for Current/PrevRealm", and "ImportPath instead of PkgPath", I see as likely to be declined because I see strong reasons / opinions in favour of keeping them as is.

**Comment by hthieu1110** [comment]
Hi @thehowl I'm currently working on refactoring to manipulate entire test context instead of using SetTestXXX function.

According to your suggestion, we can create the new Context on Gno side then we can call testing.SetContext or testing.Context.Run(f func()) to launching the test on the given context.

As I understand, the real exec context is on Go side: std.TestExecContext and we have some bindings TestSetXXX (gno) => testSetXXX => X_testSetXXX (go), so instead of having them separately, first step is to create a  X_testSetContext which allow to modify any value in the go exec context from the testing.SetContext ? Am I on the right direction ?
Or do you suggest another better way to interact with go context, gnoMachine from gno side ? Thanksss.

**Comment by thehowl** [comment]
Hey @hthieu1110,

Let's work on a `testing.SetContext` function for now, so without the `testing.Context.Run`. This should set the context for the test being executed.

Your analysis is correct, yes; however native bindings doesn't support passing in structs (and it shouldn't). Here's how I see a potential SetContext function working:

```go
type Context struct {
	IsOrigin     bool
	CurrentRealm std.Realm
	PrevRealm    std.Realm
	OrigSend     chain.Coins
	ChainID      string
	Height       uint64
	Time         time.Time
	Banker
	Logger
}

func (t *T) SetContext(ctx Context) {
	setExecContext(
		ctx.IsOrigin,
		ctx.CurrentRealmAddr,
		ctx.CurrentRealmPath,
		origSendDenoms, origSendAmounts, // like we do for origSend()
		ctx.ChainID,
		ctx.Height,
		ctx.Time.Unix(),
		ctx.Time.Nanosecond(),
		ctx.Banker,
		ctx.Logger,
	)
}

// maybe could just be std.Banker, but we can also make it an own interface.
type Banker interface { ... }

type DefaultBanker struct {
	Balances map[string]std.Coins
}

// similar for the Logger interface.
type Logger interface { ... }

type WriterLogger struct { io.Writer }

```

So:

- For most parameters, we simply have a big function with a lot of parameters we use to pass in data.
- For `Banker` and `Logger`, we'll actually want the user to be able to provide custom implementations to create mocks for the two interfaces. We should have these parameters passable to `setExecContext`, but the corresponding version in Go should use `gno.TypedValue`, so that we can then create wrappers on the actual `ExecContext` that, when called, run these methods. (This may be a bit tricky, feel free to reach out for help).
  - This is an exception in native bindings, but you can pass in any value from Gno to Go as long as the Go version uses `gno.TypedValue`.

For now, we can make this work by simply swapping out the context, as we have a mechanism that resets the context each time a test is started. This will not work in subtests; but I think we'll have to solve that in the testing system, so that we make sub-tests run separately and also with their own context; this is out of scope for your work.

Does this sound good? :)

**Comment by thehowl** [comment]
Taking a look.

> BTW, I don't get the point in file structure between gnovm/stdlibs/testing vs gnovm/tests/stdlibs/std. Should we move all testing dedicated stuffs to gnovm/stdlibs/testing as the testing context is defined in there ?

Yes, a fair question. I think we should move everything to `gnovm/tests/stdlibs` for now. Over the long term, we should likely move it on-chain, so that folks can easily build re-usable testing libraries; but let's keep it off-chain in the testing stdlibs so we can still iterate while the chain is launched.

**Comment by hthieu1110** [comment]
Hi @thehowl,

I propose that we can do the first step to move all std.TestSetXXX => t/testing.SetXXX (event std.TestIssueCoins => t.IssueCoins) then make the refactors related to Bank in a separate PR because it will not change the usages, it will change just the implementation when bridging from Gno => Go.

Could you please take a look on https://github.com/gnolang/gno/pull/3582 ?
Thanks.

**Comment by thehowl** [comment]
upon a suggestion of @moul, let's consider whether we want to actually replace all usages of `Address` to `Addr`. If there is consensus that Addr is unambiguous, clear and easier to use, then we should switch to it.

**Comment by leohhhn** [comment]
One more thing @thehowl: https://github.com/gnolang/gno/pull/4374#discussion_r2205208107
